examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts

```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
-                return;
-            }
+            // doTheMigration() is called at the start of migration and should only resolve in two cases. First, is if
+            // we complete the migration. Second, is if we disconnect during the migration process. In both cases we
+            // should re-enter the state machine and take the appropriate action (see then() block below).
 
-            const createResponse = await this.modelLoader.createDetached(acceptedVersion);
-            const migratedModel: IMigratableModel = createResponse.model;
-
-            const exportedData = await migratable.exportData();
-
-            // TODO: Is there a reasonable way to validate at proposal time whether we'll be able to get the exported
-            // data into a format that the new model can import?  If we can determine it early, then clients with old
-            // ModelLoaders can use that opportunity to dispose early and try to get new ModelLoaders.
-            let transformedData: unknown;
-            if (migratedModel.supportsDataFormat(exportedData)) {
-                // If the migrated model already supports the data format, go ahead with the migration.
-                transformedData = exportedData;
-            } else if (this.dataTransformationCallback !== undefined) {
-                // Otherwise, try using the dataTransformationCallback if provided to get the exported data into
-                // a format that we can import.
-                try {
-                    transformedData = await this.dataTransformationCallback(exportedData, migratedModel.version);
-                } catch {
-                    // TODO: This implies that the contract is to throw if the data can't be transformed, which isn't
-                    // great.  How should the dataTransformationCallback indicate failure?
+            const prepareTheMigration = async () => {
+                if (this._preparedDetachedModel !== undefined) {
+                    // If we already prepared the detached model then exit.
+                    return;
+                }
+
+                // It's possible that our modelLoader is older and doesn't understand the new acceptedVersion.
+                // Currently this fails the migration gracefully and emits an event so the app developer can know
+                // they're stuck. Ideally the app developer would find a way to acquire a new ModelLoader and move
+                // forward, or at least advise the end user to refresh the page or something.
+                // TODO: Does the app developer have everything they need to dispose gracefully when recovering with
+                // a new ModelLoader?
+                const migrationSupported = await this.modelLoader.supportsVersion(acceptedVersion);
+                if (!migrationSupported) {
                     this.emit("migrationNotSupported", acceptedVersion);
                     this._migrationP = undefined;
```

Comment:
Assignment of _migrationP seems a bit odd and hard to follow. Wouldn't it get cleared when doTheMIgration completes anyways?

Reply:
Yeah it should get cleared. Upon a second look, I also noticed we would hit the assert error below (since we are exiting for a reason other than completing the migration succesfully or disconnecting), so we probably need to update this mechanism as well.
```ts
assert(this.currentModel.migrationTool.newContainerId !== undefined,
                        "newContainerId should be defined");
```

Reply:
Feel free to resolve these comments as you make the changes.